### ISSUE TITLE
Using dvc-tracked data in huggingface datasets

### ISSUE THREAD

**Issue by dberenbaum** [iterative/dvc#9943]
From https://github.com/iterative/dvc/pull/9903:

> PS: It unfortunately does not solve my original use case of loading a file using HuggingFace datasets, it seems that `datasets` expects the URL to be fully qualified like so: `dvc::https://gitlab.com/repository/group/my-repo/my-folder/my-file.json` but this fails because `DVCFileSystem` expects the URL to point to the root of an SCM repo. Not sure how to continue from here, according to you is it an issue with DVC or HuggingFace datasets?

**Comment by pmrowla** [comment]
I think this should already be resolved

https://github.com/iterative/dvc/issues/9896#issuecomment-1720724269
https://github.com/huggingface/datasets/issues/6203#issuecomment-1720728760

**Comment by bilelomrani1** [comment]
Hi @dberenbaum and @pmrowla, thank you for your help, I was indeed using `fsspec` incorrectly, thank you @pmrowla for your help on the syntax. We managed to narrow down the problem, `datasets` is doing something weird with the URL (see https://github.com/huggingface/datasets/issues/6203#issuecomment-1721065217), but when using pure `fsspec`, I manage to load my file correctly.

**Comment by pmrowla** [comment]
Looks like there is a dvcfs bug where we aren't completely compliant with fsspec

@bilelomrani1 can you please try the linked PR

(in your hf+dvc venv:)
```
pip install 'dvc @ git+https://github.com/iterative/dvc.git@refs/pull/9951/head'
```

With that patch and the latest `datasets` release I am able to load the metrics from https://github.com/iterative/example-get-started

```python
>>> from datasets import load_dataset
load_dataset>>> load_dataset("json
KeyboardInterrupt
>>> load_dataset(
...   "json",
...   data_files="dvc://eval/metrics.json",
...   storage_options={"url": "https://github.com/iterative/example-get-started.git"},
... )
DatasetDict({
    train: Dataset({
        features: ['avg_prec', 'roc_auc'],
        num_rows: 1
    })
})
```

**Comment by daavoo** [comment]
This is worth some https://dvc.org/doc/user-guide/integrations/ page

**Comment by bilelomrani1** [comment]
@pmrowla I can confirm that it works as expected with the PR in my environment 👍

**Comment by pmrowla** [comment]
This bugfix is released in `dvc==3.21.1`
